### ISSUE TITLE
2.0.0 Extension Model/Standard

### ISSUE THREAD

**Issue by craigh** [zikula/core#2500]
This ticket is meant to encourage a discussion and (eventual) finalization of a model/standard/prototype/pattern for Core 2.0.0 Extension Bundles.

current 1.4.0 standard is [documented here](https://github.com/zikula/core/blob/1.4/src/docs/en/dev/UPGRADE-1.4.0.md#modulestructure)

Core 2.0.0 will 'evolve' this standard but is **not required** to maintain Backward Compatibility.

Easy agreement on the following:
 - Extends `\Zikula\Core\AbstractBundle` which extends `\Symfony\Component\HttpKernel\Bundle\Bundle`
   - This implies certain Bundle functionality as a Symfony Bundle.
 - PSR-4 required (PSR-0 is deprecated)
   - requires a `composer.json` file ([1.4 Module Standard](https://github.com/zikula/core/blob/1.4/src/docs/en/dev/UPGRADE-1.4.0.md#module-composerjson) & [1.4 Theme Standard](https://github.com/zikula/core/blob/1.4/src/docs/en/dev/UPGRADE-1.4.0.md#theme-composerjson))

**Comment by craigh** [comment]
Current questions in my mind:

#### Version
 - I think we can move this data to the `composer.json` file
   - how/where?
 - HookBundle Setup might be moved to [Service Container tags](http://symfony.com/doc/current/book/service_container.html#tags) with a compiler pass to gather them.

#### Installer
 - Create an Interface instead of an Abstract class.
```php
<?php
interface ExtensionInstaller
{
    public function setBundle(Bundle $bundle);
    public function install();
    public function upgrade($oldVersion);
    public function remove();
}
```
 - Most Installer implementations will also implement the `ContainerAwareInterface`, although, the container can currently be extracted from the Bundle also. Discussion here, I suppose on best way to handle this.

#### Bundle class
 - could expand this to directly set Installer class and extract Version info from composer file

#### Hooks
 - I'm wondering if it is possible to use *Service Container Tags* (see above) for all hooks instead of persisting them.

#### AdminInterface
 - Purpose: https://github.com/zikula/core/issues/2112
 - Idea: [AdminAuthInterface](https://github.com/zikula-modules/Pages/blob/master/AdminAuthInterface.php) / [AdminAuthListener](https://github.com/zikula-modules/Pages/blob/master/Listener/AdminAuthListener.php)

#### Zikula\Core\Api\AbstractApi
 - Needed? https://github.com/zikula/core/issues/2494
 - temporarily implemented in Core, but maybe isn't needed
 - a current solution for the `getLinks()` method and providing other services (like translation) to the Api
   - [Translation could be moved to a Trait](https://github.com/zikula/core/issues/2493)

**Comment by rallek** [comment]
I thought the new module structure of 1.4 are the forward layer for 2.0. Why do we get the 1.4 with all the BC problems if 2.0.0 will get a new module structure again?

Could be my missunderstanding of the current development.

**Comment by craigh** [comment]
> I thought the new module structure of 1.4 are the forward layer for 2.0

yes, but there are several unfinalized items.

**Comment by rallek** [comment]
And why not finalize them before release of 1.4? 

Ok, I also do not want to wait for 1.4. But if the modules made for 1.4 are not running at 2.0 the complete 1.4 work is more or less for nothing. Or I am wrong?

**Comment by craigh** [comment]
Yes, 1.4.x is a compatibility release, but there is no way the 2.0.0 Spec can be fully finalized until 2.0.0 is released.

See the Pages module. This is mostly 2.0.0 compliant, but there are several Legacy/deprecated items still in use that currently have no replacements. 

Ideally, a 1.4.x release will be made that is also fully Forward Compatible with 2.0.0, but it will not be 1.4.0.

**Comment by cmfcmf** [comment]
Regarding the 
> AdminAuthInterface / AdminAuthListener

This is something cool for 1.4, but will most definitely be replaced by "a proper thing" in 2.0:
http://symfony.com/doc/current/book/security.html#securing-controllers-and-other-code
```
/**
 * @Security("has_role('ROLE_ADMIN')")
 */
```

**Comment by craigh** [comment]
> but will most definitely be replaced by "a proper thing" in 2.0:

One of the 'highlights' of Zikula over the years has been the security and permissions system. I'm sure Symfony is awesome, but we should consider how we might leverage what we have in addition to symfony's features. We don't have to always replace, replace, replace. If we do that, then we can just eliminate zikula and use symfony :smile: Anyway - I just think it is a discussion worth expansion.

@Guite - maybe we can add some time on the agenda at the camp to discuss this ticket and related topics.

**Comment by Guite** [comment]
added to list of topics on last keynote slide ;-)

**Comment by craigh** [comment]
For "fun" I've created a few code demos. see:

https://github.com/zikula/core/compare/FC-features

and

https://github.com/zikula-modules/Pages/compare/alternativeInstaller

**Comment by Kaik** [comment]
Hi, quick one so permissions system and users. Same as @craigh I think our perms system is what defines Zikula and make it special. We may want to review it make it more user friendly. I have read sf docs about auth some time ago and I do not know if this is replaceable with our perms system I think it work in different layers/scopes you know more about that than me... Users I do not know yet how sf users works but I think anything will be better than our current profile/users module.

**Comment by cmfcmf** [comment]
I really didn't intend to say: "Remove all the Zikula stuff and replace it by Symfony". What I really meant was: "Review all the *cool and good working* Zikula stuff and see if we can build bridges to the Symfony pendants".

**Comment by craigh** [comment]
#### Version Capabilities

in `composer.json`
 - string keys
  - `searchable`
  - `hook_subscriber`
  - `hook_provider`
  - `admin`
  - `user`

values for each are arrays, but I think these can be standardized and made more useful. for example:

```json
  "capabilities": {
    "hook_subscriber": {"class": "Zikula\\FooModule\\Hooks\\Definition"},
    "searchable": {"class": "Zikula\\FooModule\\Helper\\SearchHelper"},
    "admin": {"route": "zikulafoomodule_admin_index"}
  }
```

**Comment by cmfcmf** [comment]
I especially like the admin route.

**Comment by cmfcmf** [comment]
There also is the question where to place these new classes and interfaces:
- in new Bundles (`Zikula\Core\Bundle\Bar`)
- in new Components, like Symfony does (`Zikula\Component\Foo`)
- directly under `Zikula\Core`

**Comment by cmfcmf** [comment]
When moving the Hooks from the version class into the composer.json file, https://github.com/zikula/core/issues/236 should also be considered.

**Comment by cmfcmf** [comment]
Also https://github.com/zikula/core/issues/249 might be taken into account - although I think we could also close that one.

**Comment by cmfcmf** [comment]
https://github.com/zikula/core/issues/411 and https://github.com/zikula/core/issues/418 should also be considered.

**Comment by cmfcmf** [comment]
Alo see https://github.com/zikula/core/issues/1286

**Comment by craigh** [comment]
I've created a SpecModule to enable further discussion

https://github.com/zikula/SpecModule

**Comment by cmfcmf** [comment]
Also, the `bootstrap.php` file of modules has to go I suppose, refs https://github.com/zikula/CoreManagerModule/issues/2.

**Comment by craigh** [comment]
the removal of `bootstrap.php` seems like one of those "refactoring for the sake of refactoring" situations - unless I am not understanding it well. Isn't the purpose the same in both cases? what is the advantage?

**Comment by craigh** [comment]
@cmfcmf - regarding `boostrap.php`: I still am not sure this can or should be removed. While it is true that some modules will have dependencies in their `composer.json` and therefore create an autoloader for itself, many modules will not need to do so and therefore may never create that.

**Comment by Guite** [comment]
Closed as completed in the 1.4 FC layer. Feel free to reopen if you disagree.
